src/regex_impl.rs

```diff
@@ -492,6 +495,112 @@ impl<W: CodeUnitWidth> Regex<W> {
+            std::fmt::Debug,
+    {
+        self.replace_impl(subject, replacement, false, extended)
+    }
+
+    pub fn replace_all<'s>(
+        &self,
+        subject: &'s [W::SubjectChar],
+        replacement: &'s [W::SubjectChar],
+        extended: bool,
+    ) -> Result<Cow<'s, [W::SubjectChar]>, Error>
+    where
+        [<W as CodeUnitWidth>::PCRE2_CHAR]: ToOwned,
+        W::PCRE2_CHAR: TryInto<W::SubjectChar>,
+        <<W as CodeUnitWidth>::PCRE2_CHAR as TryInto<<W as CodeUnitWidth>::SubjectChar>>::Error:
+            std::fmt::Debug,
+    {
+        self.replace_impl(subject, replacement, true, extended)
+    }
+
+    #[inline]
+    fn replace_impl<'s>(
+        &self,
+        subject: &'s [W::SubjectChar],
+        replacement: &'s [W::SubjectChar],
+        replace_all: bool,
+        extended: bool,
+    ) -> Result<Cow<'s, [W::SubjectChar]>, Error>
+    where
+        [<W as CodeUnitWidth>::PCRE2_CHAR]: ToOwned,
+        W::PCRE2_CHAR: TryInto<W::SubjectChar>,
+        <<W as CodeUnitWidth>::PCRE2_CHAR as TryInto<<W as CodeUnitWidth>::SubjectChar>>::Error:
+            std::fmt::Debug,
+    {
+        let mut options: u32 = 0;
+        options |= PCRE2_SUBSTITUTE_OVERFLOW_LENGTH;
+        // TODO: this should probably be configurabe from user-side
+        options |= PCRE2_SUBSTITUTE_UNSET_EMPTY;
+        if extended {
+            options |= PCRE2_SUBSTITUTE_EXTENDED;
+        }
+        if replace_all {
+            options |= PCRE2_SUBSTITUTE_GLOBAL;
+        }
+
+        // TODO: we can use MaybeUninit to avoid allocation
+        let mut capacity = 256;
+        let mut output: Vec<W::PCRE2_CHAR> = Vec::with_capacity(capacity);
+        capacity = output.capacity();
+
+        let mut rc = unsafe {
+            self.code
+                .substitute(subject, replacement, 0, options, &mut output, &mut capacity)
+        };
+
+        if let Err(e) = &rc {
+            if e.code() == PCRE2_ERROR_NOMEMORY {
+                if output.try_reserve(capacity - output.capacity()).is_err() {
+                    return Err(rc.unwrap_err());
+                }
+                capacity = output.capacity();
+                rc = unsafe {
+                    self.code.substitute(
+                        subject,
+                        replacement,
+                        0,
+                        options,
+                        &mut output,
+                        &mut capacity,
+                    )
+                };
+            }
+        }
+
+        let s = match rc? {
+            0 => Cow::Borrowed(subject),
+            _ => {
+                // +1 to account for null terminator
+                unsafe { output.set_len(capacity + 1) }; 
+
+                // this is really just a type cast
+                let x: Vec<W::SubjectChar> = output
+                    .into_iter()
+                    .map(W::PCRE2_CHAR::try_into)
```

Comment:
Can you think of any reason why this would fail? Given that all of the inputs are valid chars, I would expect anything in the output to be a valid char as well.

To avoid the long "Debug" constraint, what would you think about something like this:

```rust
// All inputs contained valid chars, so we expect all outputs to as well.
let to_char = |c: W::PCRE2_CHAR| -> W::SubjectChar {
    c.try_into()
        .unwrap_or_else(|_| panic!("all output expected to be valid chars"))
};

// this is really just a type cast
let x: Vec<W::SubjectChar> = output
    .into_iter()
    .map(to_char)
    // we don't want to return the null terminator
    .take(capacity)
    .collect();
```

also if you want to constrain `PCRE2_CHAR` in the `CodeUnitWidth` trait that's fine too, i.e. `type PCRE2_CHAR: TryInto<Self::SubjectChar>`, as that will cut down on the duplication.

Reply:
Agreed, I cannot see how it could fail except if pcre2 gives malformed output, in which case panic-ing is probably beneficial behaviour. I will also move the constraint to the trait. Will change to a literal panic over except to avoid the debug requirement.

Reply:
Also, I am fairly certain we could just as well perform an actual typecast with `Vec::from_raw_parts`, but that's a premature optimization